### PR TITLE
CDAP-6375 refactor hydrator spark

### PR DESCRIPTION
**Pull request by albertshau**
Pulling Spark related Hydrator logic into its own module in
preparation for the new artifact that will run pipelines using
Spark Streaming.

As part of this refactor, a single Spark program can now be used
to run an entire pipeline, which removes the need for running the
planner logic and local connector datasets used by the mapreduce
engine.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT4364

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/BatchPipelineSpec.java [hunk 1]
```diff
@@ -32,20 +32,27 @@
  */
 public class BatchPipelineSpec extends PipelineSpec {
   private final List<ActionSpec> endingActions;
+  private final Resources driverResources;
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/BatchPipelineSpec.java:35]: Any reason for not having `driverResources` as a part of `PipelineSpec`?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java [hunk 2]
```diff
@@ -57,27 +62,32 @@ public PipelinePhase(Map<String, Set<StageInfo>> stages, Dag dag) {
...
-        stageInfos.addAll(stages.get(pluginType));
+      Set<StageInfo> typeStages = stagesByType.get(pluginType);
+      if (typeStages != null) {
+        stageInfos.addAll(typeStages);
       }
     }
     return Collections.unmodifiableSet(stageInfos);
   }
 
+  public StageInfo getStage(String stageName) {
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java:81]: Can this method return `null`?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+package co.cask.cdap.etl.common;
+
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.InvalidEntry;
+import com.google.common.base.Function;
+
+import java.util.Collection;
+
+/**
+ * Emitter that transforms objects emitted into another type of object which can then be fetched.
```
Comment (sagarkapare) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java]: should it be `Emitter that transforms objects emitted into another type of object which can then be fetched.`?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/BatchPipelineSpec.java (hunk 1 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/BatchPipelineSpec.java:35]: it doesn't apply for etl-realtime so its only in the BatchPipelineSpec.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java (hunk 2 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelinePhase.java:81]: yep, will mark it.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java (hunk 3 above)
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformingEmitter.java]: thats what I wrote, but can try to make it clearer.

File: cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java [hunk 4]
```diff
@@ -101,6 +107,29 @@ protected void configure() {
...
+    }
+
+    PipelinePlanner planner;
+    if (useSpark) {
+      // if the pipeline uses spark, we don't need to break the pipeline up into phases, we can just have
+      // a single phase.
+      planner = new PipelinePlanner(supportedPluginTypes, ImmutableSet.<String>of(), ImmutableSet.<String>of());
+    } else {
+      planner = new PipelinePlanner(supportedPluginTypes,
+                                    ImmutableSet.of(BatchAggregator.PLUGIN_TYPE),
```
Comment (CuriousVini) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: Missed `BatchJoiner.PLUGIN_TYPE`? :(

Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/SmartWorkflow.java]: oh. yes good catch.

**Comment by CuriousVini** [comment]
LGTM

**Comment by albertshau** [comment]
changes were non-code related. build was green so merging.

**Pull request merged**